geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/ClientRegistrationEventQueueManagerTest.java

```diff
@@ -196,4 +197,51 @@ public void addAndDrainQueueContentionTest() throws ExecutionException, Interrup
 
     assertThat(clientRegistrationEventQueue.isEmpty()).isTrue();
   }
+
+  @Test
+  public void twoThreadsRegisteringSameClientNoEventsLost()
+      throws ExecutionException, InterruptedException {
+    ClientRegistrationEventQueueManager clientRegistrationEventQueueManager =
+        new ClientRegistrationEventQueueManager();
+
+    InternalCacheEvent internalCacheEvent = mock(InternalCacheEvent.class);
+    when(internalCacheEvent.getRegion()).thenReturn(mock(LocalRegion.class));
+
+    Conflatable conflatable = mock(Conflatable.class);
+    Set<ClientProxyMembershipID> filterClientIDs = new HashSet<>();
+    CacheClientNotifier cacheClientNotifier = mock(CacheClientNotifier.class);
+    ClientProxyMembershipID clientProxyMembershipID = mock(ClientProxyMembershipID.class);
+
+    ClientRegistrationEventQueueManager.ClientRegistrationEventQueue clientRegistrationEventQueue =
+        clientRegistrationEventQueueManager.create(clientProxyMembershipID,
+            new ConcurrentLinkedQueue<>(), new ReentrantReadWriteLock());
+
+    for (int registrationIterations = 0; registrationIterations < 1000; ++registrationIterations) {
+      Runnable clientRegistrationSimulation = () -> {
+        for (int numAdds = 0; numAdds < getRandomNumberOfAdds(); ++numAdds) {
+          // In thread one, we add events to the queue
+          clientRegistrationEventQueueManager
+              .add(internalCacheEvent, conflatable, filterClientIDs, cacheClientNotifier);
+        }
+        // In thread two, we drain events from the queue
+        clientRegistrationEventQueueManager.drain(clientProxyMembershipID, cacheClientNotifier);
+      };
+
+      CompletableFuture<Void> registrationFutureOne =
+          CompletableFuture.runAsync(clientRegistrationSimulation);
+      CompletableFuture<Void> registrationFutureTwo =
+          CompletableFuture.runAsync(clientRegistrationSimulation);
+
+      CompletableFuture.allOf(registrationFutureOne, registrationFutureTwo).get();
+
+      assertThat(clientRegistrationEventQueue.isEmpty()).isTrue();
+    }
+  }
+
+  /*
+   * This helps to create contention between registration threads during the drain phase
+   */
+  private static int getRandomNumberOfAdds() {
+    return new Random().nextInt(((10000 - 50000) + 1) + 50000);
```

Comment:
This equals `10_001`. So you'll get ints in the range [`0`,`10_000`] (inclusive).

Where do these constants come from?

When this fn returns `0` the `for` loop  at line 221 is never entered.

Reply:
Danke - the math is all bunk here.